libcxx/include/__format/format_functions.h

```diff
@@ -251,7 +251,7 @@ __handle_replacement_field(_Iterator __begin, _Iterator __end, _ParseCtx& __pars
   if (__r.__last == __end)
     std::__throw_format_error("The argument index should end with a ':' or a '}'");
 
-  bool __parse = *__r.__last == _CharT(':');
+  bool __parse = __parse_ctx.__should_parse() = *__r.__last == _CharT(':');
```

Comment:
I did some more investigation, this seems to be the wrong approach. It solves
`std::println("{}>42", std::thread::id{});`, but fails with `std::println("{:}>42", std::thread::id{});`.

I did a bit more testing on [godbolt](https://godbolt.org/z/shfbaroj5) and other cases for libc++ fail too.

We should take a different approach to fix this
https://github.com/llvm/llvm-project/blob/main/libcxx/include/__format/parser_std_format_spec.h#L580
```
    if (__use_range_fill && (__fill == _CharT('{') || __fill == _CharT('}') || __fill == _CharT(':')))
      std::__throw_format_error("The fill option contains an invalid value");
    else if (__fill == _CharT('{') || __fill == _CharT('}'))
      std::__throw_format_error("The fill option contains an invalid value");
```
here we should not test for '}' instead 

https://github.com/llvm/llvm-project/blob/main/libcxx/include/__format/parser_std_format_spec.h#L609
```
 __validate_fill_character(*__begin, __use_range_fill);
```
should first test whether the fill character is `_CharT('}')` and return `false` when this condition is true.

I would like to see tests for the cases this fixes.

Reply:
After reading your comment and the Standard, I'm thinking maybe doing an early return just before calling `__parse_fill_align` is the right thing to do.
https://github.com/llvm/llvm-project/blob/b45de48be24695b613f48ed21bb35f844454193b/libcxx/include/__format/parser_std_format_spec.h#L355-L361
```diff
-     if (__begin == __end)
+     if (__begin == __end || *__begin == _CharT('}'))
```

According to [[format.string.general]/1](https://eel.is/c++draft/format.string.general#1), `}` is definitely not the start of `format-spec` so when we see a `}` at the start of `__parse_fill_align`, that must belong to `replacement-field` and the `format-spec` is empty. Specialized formatters (`__validate_fill_character` that is) shouldn't even have a say about that `}`, [[format.string.general]/5](https://eel.is/c++draft/format.string.general#5).

For example, we can parse `{:}<}` as having an empty `format-spec` followed by an invalid escape sequence rather than having an invalid `fill` followed by an `align` and a closing brace. The ill-fomed-ness doesn't change but the exception.what() does, so we need to fix the tests checking this if we are going this route.

All library types' format specifications seem to be optional and maybe that's designed like so to allow this way of parsing.

Reply:
> After reading your comment and the Standard, I'm thinking maybe doing an early return just before calling __parse_fill_align is the right thing to do.

Yes testing for `_CharT('}')` before calling `__parse_fill_align` sounds like a cleaner solution my proposed solution.

> Specialized formatters (__validate_fill_character that is) shouldn't even have a say about that }, 

Correct it can still call out `{` since this is always invalid. Maybe we need to investigate ':' for range formatters in more detail. "{:::<<}" is probably rejected. I expect it should be an underlying formatter with left-alignment and `<` as fill character. (I have not tested that.) Maybe it would be good to investigate that for a follow-up patch.

> For example, we can parse {:}<} as having an empty format-spec followed by an invalid escape sequence rather than having an invalid fill followed by an align and a closing brace. The ill-fomed-ness doesn't change but the exception.what() does, so we need to fix the tests checking this if we are going this route.

Correct that is exactly what we should do.

> Does this concludes what you mean?

Yes.